engine/tools/src/texconvert/main.cpp

```diff
@@ -0,0 +1,262 @@
+                params.m_Verbose = 1;
+            else if (CMP_ARG_1_OP("color-space"))
+                params.m_ColorSpace = GetArgTypeValue(argv[++i], g_cs_lut, params.m_ColorSpace);
+            else if (CMP_ARG_1_OP("compression-type"))
+                params.m_CompressionType = GetArgTypeValue(argv[++i], g_ct_lut, params.m_CompressionType);
+            else if (CMP_ARG_1_OP("compression-level"))
+                params.m_CompressionLevel = GetArgTypeValue(argv[++i], g_cl_lut, params.m_CompressionLevel);
+
+            #undef CMP_ARG_1_OP
+            #undef CMP_ARG
+        }
+    }
+}
+
+int DoEncode(EncodeParams params)
+{
+    int x,y,n;
+    unsigned char *data = stbi_load(params.m_PathIn, &x, &y, &n, 0);
+
+    dmTexc::HTexture tex = dmTexc::Create(params.m_PathIn, x, y,
+        GetPixelFormat(n), params.m_ColorSpace, params.m_CompressionType, data);
+
+    if (params.m_PremultiplyAlpha && !dmTexc::PreMultiplyAlpha(tex))
+    {
+        printf("Unable to premultiply alpha\n");
+        return -1;
+    }
+
+    if (params.m_FlipY && !dmTexc::Flip(tex, dmTexc::FLIP_AXIS_Y))
+    {
+        printf("Unable to flip Y\n");
+        return -1;
+    }
+
+    if (params.m_FlipX && !dmTexc::Flip(tex, dmTexc::FLIP_AXIS_X))
+    {
+        printf("Unable to flip Y\n");
+        return -1;
+    }
+
+    if (!dmTexc::Encode(tex, dmTexc::PF_RGBA_BC3, params.m_ColorSpace,
+                params.m_CompressionLevel, params.m_CompressionType, params.m_MipMaps, 4))
+    {
+        printf("Unable to encode texture data\n");
+        return -1;
+    }
+
+    uint32_t out_data_size = dmTexc::GetTotalDataSize(tex);
+    void* out_data         = malloc(out_data_size);
+    dmTexc::GetData(tex, out_data, out_data_size);
+
+    dmTexc::Destroy(tex);
+
+    stbi_image_free(data);
+
+    FILE* f = fopen(params.m_PathOut, "w");
+
+    fwrite(out_data, out_data_size, 1, f);
+    fclose(f);
+    return 0;
+}
+
+void ShowHelp()
+{
+#define PRINT_ARG_LIST(lst) \
+    { \
+        for(int i=0; lst[i].m_Name != 0; i++) \
+            printf("    %s\n", lst[i].m_Name); \
+    }
+    printf("Usage: texconvert <input-file> <output-file> [options]\n");
+    printf("Options:\n");
+    printf("  --premultiply-alpha         : Use premultiply alpha\n");
+    printf("  --flip-x                    : Flip image on X axis\n");
+    printf("  --flip-y                    : Flip image on Y axis\n");
+    printf("  --verbose                   : Verbose output\n");
+    printf("  --color-space <color-space> : Sets the color space, defaults to 'SRGB'. Supported values are:\n");
+    PRINT_ARG_LIST(g_cs_lut);
+    printf("  --compression-type          : Sets the compression type, defaults to 'DEFAULT'. Supported values are:\n");
+    PRINT_ARG_LIST(g_ct_lut);
+    printf("  --compression-level         : Sets the compression level, defaults to 'NORMAL'. Supported values are:\n");
+    PRINT_ARG_LIST(g_cl_lut);
+#undef PRINT_ARG_LIST
+}
+
+int main(int argc, char const *argv[])
```

Comment:
I created this cmd line tool so that it's easy to use our internal encoder to convert between common image formats into basis files. We use the Texc library in bob as well, which we use in the TextureGenerator to generate test content for the engine and we *could* restructure that to do something similar like this tool, but the benefit here is:
* users can use the tool to produce runtime data with correct formats/versions we support in the engine and don't need to deal with java (class paths etc)
* the java tooling requires texture profiles to output compressed and transcoded formats
* this tool is more isolated which means it's easier to distribute in a game team

I guess we could move the functionality to the java class through waf_texc.py or something like that, but I personally would prefer to keep using this small tool instead of our pipeline